### PR TITLE
ARROW-8659: [Rust] ListBuilder allocate with_capacity

### PR DESCRIPTION
**Pull request by tustvold**
Both ListBuilder and FixedSizeListBuilder accept a values_builder as a constructor argument and then set the capacity of their internal builders based off the length of this values_builder. Unfortunately at construction time this values_builder is normally empty, and consequently programs spend an unnecessary amount of time reallocating memory.

This PR adds new constructor methods that allow specifying the desired capacity upfront.

### CODE REVIEW

File: rust/parquet/src/arrow/converter.rs [hunk 1]
```diff
@@ -128,7 +128,10 @@ pub struct Utf8ArrayConverter {}
 
 impl Converter<Vec<Option<ByteArray>>, StringArray> for Utf8ArrayConverter {
     fn convert(source: Vec<Option<ByteArray>>) -> Result<StringArray> {
-        let mut builder = StringBuilder::new(source.len());
+        let mut builder = StringBuilder::with_capacity(
+            source.len(),
+            source.len() * std::mem::size_of::<ByteArray>(),
```
Comment (tustvold) [inline rust/parquet/src/arrow/converter.rs:133]: This might be a touch on the pessimistic side, but appears to be the same approximation made in `ComplexObjectArrayReader`. It would be nice to somehow feed the data from `ComplexObjectArrayReader` through to the converter, but this was a non-trivial change so I avoided it.

Comment (sunchao) [reply rust/parquet/src/arrow/converter.rs:133]: Yeah this could be overly pessimistic. A more accurate way is to calculate the sum of all lens. It may be worth it.

File: rust/arrow/src/array/builder.rs [hunk 2]
```diff
@@ -527,11 +527,18 @@ pub struct ListBuilder<T: ArrayBuilder> {
 impl<T: ArrayBuilder> ListBuilder<T> {
     /// Creates a new `ListArrayBuilder` from a given values array builder
     pub fn new(values_builder: T) -> Self {
-        let mut offsets_builder = Int32BufferBuilder::new(values_builder.len() + 1);
+        let capacity = values_builder.len();
+        Self::with_capacity(values_builder, capacity)
+    }
+
+    /// Creates a new `ListArrayBuilder` from a given values array builder
+    /// `append` may be called up to `capacity` times without triggering reallocation
```
Comment (sunchao) [inline rust/arrow/src/array/builder.rs:535]: nit: maybe rephrase this to: "`capacity` is the number of items to pre-allocate space for in offset buffer of this builder"? similarly below.

**Review by sunchao — COMMENTED** [review]
LGTM just a few nits :)

File: rust/arrow/src/array/builder.rs (hunk 2 above)
Comment (tustvold) [reply rust/arrow/src/array/builder.rs]: I went with something I hope was in the spirit of what you wanted, as technically it allocates for an extra offset and also pre-allocates the bitmap builder. Let me know if you'd prefer I stuck with your exact wording

File: rust/parquet/src/arrow/converter.rs (hunk 1 above)
Comment (tustvold) [reply rust/parquet/src/arrow/converter.rs]: Having revisited this, not only is it pessimistic - it's plain wrong :). The ByteArray type here is just a wrapper pointer struct and its size in memory has no semantic value. This is fixed in the latest version which computes the correct array capacity.

**Review by sunchao — APPROVED** [review]
Nice! LGTM

**Comment by sunchao** [comment]
Merged. Thanks @tustvold !
